GitUI/CommandsDialogs/CustomDiffMergeTool.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System.Collections.Generic;
+using System.ComponentModel;
+using System.Drawing;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using GitCommands;
+
+namespace GitUI
+{
+    public class CustomDiffMergeTool
+    {
+        public CustomDiffMergeTool(ToolStripMenuItem menuItem, System.EventHandler click)
+        {
+            _menuItem = menuItem;
+            _click = click;
+        }
+
+        private ToolStripMenuItem _menuItem { get; set; }
+        private System.EventHandler _click;
+
+        private static SemaphoreSlim _diffToolMutex = new(1);
+        private static SemaphoreSlim _mergeToolMutex = new(1);
+        private static IEnumerable<string> _diffTools = null;
+        private static IEnumerable<string> _mergeTools = null;
+
+        /// <summary>
+        /// Clear the existing caches
+        /// </summary>
+        public static void Clear()
+        {
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await _diffToolMutex.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    _diffTools = null;
+                }
+                finally
+                {
+                    _diffToolMutex.Release();
+                }
+
+                await _mergeToolMutex.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    _mergeTools = null;
+                }
+                finally
+                {
+                    _mergeToolMutex.Release();
+                }
+            }).FileAndForget();
+        }
+
+        /// <summary>
+        /// Load the availble DiffMerge tools and apply to the menus
+        /// </summary>
+        /// <param name="module">The Git module</param>
+        /// <param name="menus">The menus to update</param>
+        /// <param name="components">The calling Form components, to dispose correctly</param>
+        /// <param name="isDiff">True if diff, false if merge</param>
+        /// <param name="delay">The delay before starting the operation</param>
+        public static void LoadCustomDiffMergeTools(GitModule module, IList<CustomDiffMergeTool> menus, IContainer components, bool isDiff, int delay = 1000)
```

Comment:
Why do you feel we need to re-run this method instead of loading all available diff/merge tools once at the app startup? 
It isn't a common scenario where a user constantly installs/uninstalls diff tools. Loading the tools only once we will greatly simplify the code and associated plumbing. We also will need to build and maintain a single instance of `ContextMenuStrip`, and we won't need to maintain its lifetime.

Reply:
`LoadCustomDiffMergeTools` is called at startup of the form. The run is slightly delayed to not compete with CPU for git-log.
Each form need to update each of the menu items that handle difftools (so RevDiff has five menu items).

There is a minimal extra plumbing to reload. This is not a major functionality, I just felt that a user expected the list was refreshed after closing Settings, when `Clear` is called and the difftools are reloaded. (This ignores that it is possible to add diff/mergetools to repo settings).